utils/s3/client.hh

```diff
@@ -30,10 +31,43 @@ namespace s3 {
 
 using s3_clock = std::chrono::steady_clock;
 
-struct range {
-    uint64_t off;
-    size_t len;
+class range {
+    friend struct fmt::formatter<range>;
+
+public:
+    range() noexcept = delete;
+    constexpr range(uint64_t offset, uint64_t len) noexcept : _offset(offset), _length(len) {
+        uint64_t t;
+        SCYLLA_ASSERT(!__builtin_add_overflow(_offset, len, &t));
+        SCYLLA_ASSERT(_offset + len <= _max_object_size);
+    }
+    explicit constexpr range(uint64_t offset) noexcept : _offset(offset), _length(std::nullopt) {}
+    constexpr range& operator+=(uint64_t offset) noexcept {
+        uint64_t t;
+        SCYLLA_ASSERT(!__builtin_add_overflow(_offset, offset, &t));
+        SCYLLA_ASSERT(_offset + offset <= _max_object_size);
+
+        _offset += offset;
+        if (_length) {
+            SCYLLA_ASSERT(!__builtin_sub_overflow(_length.value(), offset, &t));
+            _length.value() -= offset;
+        }
+        return *this;
+    }
+    bool operator==(const range& other) const noexcept { return std::tie(_offset, _length) == std::tie(other._offset, other._length); }
+    [[nodiscard]] sstring to_header_string() const noexcept {
+        return fmt::format("bytes={}-{}", _offset, _length ? fmt::format("{}", _offset + *_length - 1) : "");
+    }
+    [[nodiscard]] uint64_t offset() const noexcept { return _offset; }
+    [[nodiscard]] std::optional<uint64_t> length() const noexcept { return _length; }
+
+private:
+    // 5TiB is the largest object size supported by S3
+    static constexpr uint64_t _max_object_size{5_TiB};
+    uint64_t _offset;
+    std::optional<uint64_t> _length;
```

Comment:
Optional length makes checks in client code mode complicated. Why can't disengaged optional be replaced with _max_object_size instead?